### ISSUE TITLE
Mask image for test

### ISSUE THREAD

**Issue by naoki7090624** [SayedNadim/Global-and-Local-Attention-Based-Free-Form-Image-Inpainting#4]
Thank you for sharing your great job!
I try to run this model on my dataset without mask images. I found that `train.py` works, but `test_single.py` needs a mask image to output. Is it possible to work without the mask image?

**Comment by SayedNadim** [comment]
Hi, 
Thanks for the question. Currently, the model only supports test input and test mask. If you need to generate mask, you can use the following code:

`
import numpy as np
import cv2
import torch
import torchvision.utils as vutils

class Masks():

    @staticmethod
    def get_ff_mask(h, w, num_v=None):
        # Source: Generative Inpainting https://github.com/JiahuiYu/generative_inpainting

        mask = np.zeros((h, w))
        if num_v is None:
            num_v = 15 + np.random.randint(9)  # 5

        for i in range(num_v):
            start_x = np.random.randint(w)
            start_y = np.random.randint(h)
            for j in range(1 + np.random.randint(5)):
                angle = 0.01 + np.random.randint(4.0)
                if i % 2 == 0:
                    angle = 2 * 3.1415926 - angle
                length = 10 + np.random.randint(60)  # 40
                brush_w = 10 + np.random.randint(15)  # 10
                end_x = (start_x + length * np.sin(angle)).astype(np.int32)
                end_y = (start_y + length * np.cos(angle)).astype(np.int32)

                cv2.line(mask, (start_y, start_x), (end_y, end_x), 1.0, brush_w)
                start_x, start_y = end_x, end_y

        return (mask > 0).astype(np.float32)

def mask_image(height, width):
    mask = Masks.get_ff_mask(height, width)
    mask = torch.from_numpy(np.asarray(mask)).unsqueeze(0).unsqueeze(1).float()
    ones = torch.ones(1, 1, height, width)
    mask = ones * mask
    return mask


mask = mask_image(256,256)
vutils.save_image(mask,'mask.png')
`
Cheers!

**Comment by SayedNadim** [comment]
Hi,
Currently, the model only supports a test image with a test mask. If you need to generate mask, you can use the 'mask_image' function in 'mask.py' under 'model' directory. 
You can refer to [this](https://github.com/SayedNadim/Global-and-Local-Attention-Based-Free-Form-Image-Inpainting/blob/ba1fbae34c120ebc605472cce63d8e8dabc83880/model/mask.py#L38).
Let me know if you need any help.
Cheers!

**Comment by naoki7090624** [comment]
Thank you for your quick response.
When I input the mask image that value has 0 and 1, the result was same as input image. Then I generate the mask image as an RGB image (=3 channel) with value of 0 and 255. The results was different from input images, but with some artifacts. 
What value should I input for the mask image?
I'd like to see if I failed to train my model or if there is a problem with the mask image.

**Comment by SayedNadim** [comment]
Hi,
Sorry for the late reply.  The mask values should be binary (i.e. 0 or 1). That is because when we create corrupt image, we multiply mask with image i.e. `x*(1. - mask)`. `(1. - mask)` means 0 for holes and 1 for the non-hole area. So, after multiplication, the corrupted image has a value range of  0-255.
If you want to use RGB mask, I would suggest you to normalize the mask values in 0-1 and then try.
Let me know if you need any help.
Cheers!

**Comment by naoki7090624** [comment]
Thank you for the advice.
If I use the RGB mask image that normalized 0 to 1 for ```mask = default_loader(args.mask)```,  the result of ```x = x * (1. - mask)```  will have a value range of -1 to 1 as bellow. If I use the 0-255 mask image for input, ```x``` will have the value range of 0-255. But the results has artifact. It may be overlearning at my training, but I want to make sure the result using the right mask image. I would be grateful if you could share a mask image or datasets.
```
tensor([[[[ 0.1451,  0.1451,  0.1686,  ...,  0.2305,  0.2227,  0.2227],
          [ 0.0902,  0.1137,  0.1529,  ...,  0.1992,  0.2148,  0.2305],
          [ 0.0745,  0.1059,  0.1686,  ...,  0.1758,  0.1992,  0.2227],
          ...,
          [-0.1608, -0.1608, -0.1608,  ..., -0.5843, -0.4039, -0.3490],
          [-0.1529, -0.1686, -0.1686,  ..., -0.5843, -0.4196, -0.4275],
          [-0.1529, -0.1608, -0.1765,  ..., -0.5608, -0.4824, -0.5294]],

         [[ 0.2235,  0.2235,  0.2471,  ...,  0.3242,  0.3164,  0.3164],
          [ 0.1686,  0.1922,  0.2314,  ...,  0.2930,  0.3086,  0.3242],
          [ 0.1608,  0.1922,  0.2549,  ...,  0.2852,  0.3086,  0.3320],
          ...,
          [ 0.0980,  0.0980,  0.0980,  ..., -0.5451, -0.3647, -0.3098],
          [ 0.0902,  0.0902,  0.0902,  ..., -0.5373, -0.3961, -0.4039],
          [ 0.0902,  0.0824,  0.0745,  ..., -0.5137, -0.4588, -0.5059]],

         [[ 0.2157,  0.2157,  0.2392,  ...,  0.5117,  0.5039,  0.5039],
          [ 0.1765,  0.2000,  0.2392,  ...,  0.4961,  0.5117,  0.5273],
          [ 0.1922,  0.2235,  0.2706,  ...,  0.4961,  0.5195,  0.5430],
          ...,
          [ 0.2157,  0.2157,  0.2314,  ..., -0.6000, -0.4196, -0.3647],
          [ 0.2314,  0.2392,  0.2392,  ..., -0.6157, -0.4667, -0.4745],
          [ 0.2471,  0.2392,  0.2392,  ..., -0.6078, -0.5294, -0.5765]]]])
```

**Comment by SayedNadim** [comment]
Hi,
I have created a gist for your mask generation. Please refer to [this](https://gist.github.com/SayedNadim/0f4ebd721dcd29a8b602d04a3510dd61). Let me know if it helps.
Cheers!

**Comment by naoki7090624** [comment]
I am grateful for your support.
I used mask generated by your script and the results have artifact. It seems that it was overlearning on my model.
I try to train again. Thanks for your kind response.

**Comment by SayedNadim** [comment]
Hi,
Did you try with the updated code? I have made some minor changes in the attention layer.  Let me know if you still have artifacts.
Cheers!

**Comment by naoki7090624** [comment]
Yes. I generated the mask image with ```ff_mask_generation.py```. In fact, I did this work with the not so famous datasets. The training images are only 6,000 so the results have  artifacts due to overlearning. Then, I try to train with Places2.

**Comment by SayedNadim** [comment]
Ok. You can try augmentation to increase your dataset. You can introduce random crops, random flips and color jitter to increase number of images.
If you think your issue is resolved, please close the issue.
Cheers!
